### ISSUE TITLE
Status inconsistencies between deployment and its pods

### ISSUE THREAD

**Issue by kayrus** [kubernetes/kubernetes#82405]
**What happened**:

I face issues with the deployment status. Its pods are ready and have **Running** status, however the deployment status doesn't show readiness.

```
$ kubectl get pods -l role=server
NAME                             READY   STATUS    RESTARTS   AGE
server-bc9c5c7b8-2vv6f   3/3     Running   0          44h
server-bc9c5c7b8-6m9j5   3/3     Running   2          46h
server-bc9c5c7b8-76zdb   3/3     Running   3          44h
$ kubectl get deployment server
NAME             READY   UP-TO-DATE   AVAILABLE   AGE
server   2/3     3            2           21d
```

**What you expected to happen**:

I expect the see the **Ready 3/3**  deployment status.

**How to reproduce it (as minimally and precisely as possible)**:

This happens, when a kubelet reestablishes a connection to a kube-apiserver. Not always, but I'm able to reproduce the issue with 50% chance.

Some further debugging showed that the pod status cache map stuck with `Ready True` in `pkg/kubelet/status/status_manager.go` as an old and a new value, therefore reconciliation is not triggered.

Cache map in `pkg/kubelet/config/config.go` stuck with `Ready False` for both old and new pod statuses, reconciliation is not triggered as well.

For some reason reconciliation loops don't merge these two values and they persist until you restart kubelet. Still trying to understand what exactly is wrong (probably one line fix is needed, where pointer is used instead of DeepCopy, probably some lack of mutex lock)

**Environment**:
- Kubernetes version (use `kubectl version`): 1.15.4
- Cloud provider or hardware configuration: openstack
- OS (e.g: `cat /etc/os-release`): coreos stable

@kubernetes/sig-scheduling

**Comment by kayrus** [comment]
@kubernetes/sig-scheduling
/sig scheduling

**Comment by Joseph-Irving** [comment]
Can you reproduce this issue on a more recent version of Kubernetes? v1.5.0 is a few years old now and is no longer supported, the current supported versions are v1.13, v1.14 and v1.15

**Comment by kayrus** [comment]
@Joseph-Irving sorry, it's a typo, the version is 1.15

**Comment by Joseph-Irving** [comment]
can you show the full output of the deployment, `kubectl get deployment server -o yaml` ?

**Comment by kayrus** [comment]
@Joseph-Irving here is the output for another deployment with the same problem:
https://gist.github.com/kayrus/8a5956126f08fe974c91f5fb64e2e6a4

**Comment by Joseph-Irving** [comment]
Interesting there are a few things that don't look write there, Pod ready is `false` [here](https://gist.github.com/kayrus/8a5956126f08fe974c91f5fb64e2e6a4#file-pod-yaml-L95) but container is ready [here](https://gist.github.com/kayrus/8a5956126f08fe974c91f5fb64e2e6a4#file-pod-yaml-L111) 

Deployment says it has 1 unavailable replica [here](https://gist.github.com/kayrus/8a5956126f08fe974c91f5fb64e2e6a4#file-deployment-yaml-L103) 

The replicaset seems to have an incomplete [status field](https://gist.github.com/kayrus/8a5956126f08fe974c91f5fb64e2e6a4#file-replicaset-sh-L96). I would expect it to have `readyReplicas` and `availableReplicas`

**Comment by kayrus** [comment]
@Joseph-Irving do you have a clue where to dig further?

**Comment by Joseph-Irving** [comment]
So the reason the Deployment/Replicaset don't appear to have the correct status is because the Pod has the condition `Ready` set to `False`. 
However this should be true, https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/generate.go#L94 `PodReady` is set to `True` if `ContainersReady` is `True` and all your readiness gates and ready (you're not using any pod readiness gates). So in your case `PodReady` should be true. 

`updateStatusInternal` should go an update the pod status on the api server https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/status_manager.go#L362, so I would have a look in your kubelet log, specifically looking for things status_manager.go to see if there are any errors when attempting to update the status.

**Comment by kayrus** [comment]
@Joseph-Irving I was able to reproduce this case by using iptables rule:

```
iptables -I OUTPUT -d kubeapiserver -j REJECT --reject-with icmp-port-unreachable
```

see the animation attached.
![kubelet_status_bug](https://user-images.githubusercontent.com/4047483/68955738-fdba8580-07c6-11ea-951f-32733b64c436.gif)

**Comment by kayrus** [comment]
I added more debug into the https://github.com/kubernetes/kubernetes/blob/7a5929da02f9fbdb9d63b5335f159462cc553020/pkg/kubelet/status/status_manager.go#L497 and it appeares that the `status.status` contains `Type:Ready Status:True`, when the actual kube-apiserver status is `Type:Ready Status:False`.

Then I added more debug into the https://github.com/kubernetes/kubernetes/blob/7a5929da02f9fbdb9d63b5335f159462cc553020/pkg/kubelet/config/config.go#L448 and it appeared that both `existing` and `ref` contain `Ready False`.

So far I suspect this func: https://github.com/kubernetes/kubernetes/blob/7a5929da02f9fbdb9d63b5335f159462cc553020/pkg/kubelet/config/config.go#L252

**Comment by kayrus** [comment]
@databus23 

the patch below, adapted to  v1.15.x from the #84951, PR solved the issue. Tested multiple times.

```diff
diff --git a/pkg/kubelet/kubelet_getters.go b/pkg/kubelet/kubelet_getters.go
index 94cb68fa2fe..aa8c627f9c2 100644
--- a/pkg/kubelet/kubelet_getters.go
+++ b/pkg/kubelet/kubelet_getters.go
@@ -30,6 +30,7 @@ import (
        "k8s.io/kubernetes/pkg/kubelet/cm"
        "k8s.io/kubernetes/pkg/kubelet/config"
        kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+       kubepod "k8s.io/kubernetes/pkg/kubelet/pod"
        "k8s.io/kubernetes/pkg/util/mount"
        utilnode "k8s.io/kubernetes/pkg/util/node"
        utilpath "k8s.io/utils/path"
@@ -171,8 +172,11 @@ func (kl *Kubelet) GetPods() []*v1.Pod {
        // a kubelet running without apiserver requires an additional
        // update of the static pod status. See #57106
        for _, p := range pods {
-               if status, ok := kl.statusManager.GetPodStatus(p.UID); ok {
-                       p.Status = status
+               if kubepod.IsStaticPod(p) {
+                       if status, ok := kl.statusManager.GetPodStatus(p.UID); ok {
+                               klog.V(2).Infof("status for pod %v updated to %v", p.Name, status)
+                               p.Status = status
+                       }
                }
        }
        return pods
```

**Comment by kayrus** [comment]
fixed in k8s 1.15.8 (1.15.9)
